LoRaEngine/LoraKeysManagerFacade/IoTCentralImp/DeviceTwin.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoraKeysManagerFacade.IoTCentralImp
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public class DeviceTwin : IDeviceTwin
+    {
+        private readonly DesiredProperties properties;
+
+        public DeviceTwin(string deviceName, DesiredProperties properties)
+        {
+            this.DeviceId = deviceName;
+            this.properties = properties;
+        }
+
+        public string DeviceId { get; }
+
+        public string GetDevAddr()
+        {
+            return this.properties.DevAddr;
+        }
+
+        public string GetGatewayID()
+        {
+            return this.properties.GatewayID;
+        }
+
+        public DateTime GetLastUpdated()
+        {
+            if (!this.properties.AdditionalData.TryGetValue("$metadata", out var metadata))
+            {
+                return DateTime.MinValue;
+            }
+
+            return new List<DateTime>()
+            {
+                metadata?["NwkSKey"]?["lastUpdateTime"]?.ToObject<DateTime>() ?? DateTime.MinValue,
+                metadata?["DevAddr"]?["lastUpdateTime"]?.ToObject<DateTime>() ?? DateTime.MinValue,
+                metadata?["GatewayID"]?["lastUpdateTime"]?.ToObject<DateTime>() ?? DateTime.MinValue
+            }
+            .OrderByDescending(x => x)
+            .First();
```

Comment:
The duplication of each time lookup across keys can be reduced as follow:

```suggestion
            var updateTimes =
                from x in new[] { "NwkSKey", "DevAddr", "GatewayID" }
                select metadata?[x]?["lastUpdateTime"]?.ToObject<DateTime>() ?? DateTime.MinValue into dt
                orderby dt descending
                select dt;

            return updateTimes.First();
```